testing processor (#776)
aptos-labs/developer-docs: apps/nextra/pages/en/build/indexer/indexer-sdk/documentation/advanced-tutorials/processor-test.mdx

# Overview
### What Is a Processor?
A processor is a core component of the Aptos Indexer that handles blockchain transaction processing. It validates, transforms, and stores transactions into a database, enabling downstream applications like analytics, indexing, and querying. Testing the processor ensures that all transactions are correctly handled, maintaining data accuracy and consistency.

### What Are We Testing With This?

- **Transaction correctness**: Ensure that each transaction is processed and stored accurately.
- **Schema consistency**: Verify that the database schema is correctly set up and maintained throughout the tests.

### General Flow of how Processor Testing Works

1. Prepare testing transactions (refer to prior documentations).
2. Update dependencies as needed.
3. Import new transactions.
4. Write test cases.
5. Generate expected database output and validate.
6. Merge.

## Prerequisites
<Callout>
    Key Considerations:
 - Each test runs in an isolated environment using a PostgreSQL container to prevent interference.
 - Proper handling of versions ensures transactions are processed and validated in the correct order.

>  - Validation logic must detect changes or issues by comparing processor output with the expected baseline.

Reviewer: it's unclear which part of the tutorial these key considerations are related to. maybe we add them in to the part of the tutorial that makes sense?

</Callout>

1. Ensure Docker is running for PostgreSQL container support.
    - Set up docker engine/daemon on your machine
    - Start Docker if it's not running
2. Identify the transactions to test.
   - Use imported transactions or write your own custom Move scripts to generate test transactions. Refer to Importing Transaction Guide and Generating Transaction using Move Script Guide for detailed instructions.
3. Import necessary modules, see example:
    ```rust

>     use aptos_indexer_testing_framework::{

Reviewer: we should add a pre-reqs section that tells them to add aptos_indexer_testing_framework to their Cargo.toml

        database::{PostgresTestDatabase, TestDatabase},
        sdk_test_context::SdkTestContext,
    };
    ```

## Steps to Write a Test

### 1. Set Up the Test Environment
Before setting up the test environment, it’s important to understand the configurations being used in this step:

**What Are These Configurations?**

`generate_file_flag`

> - This flag determines whether to run the test in a "diff mode" or not.
> In "diff mode," the system will compare the actual output from the processor with the expected output and highlight differences.
> Use this mode when validating changes or debugging discrepancies.

Reviewer: ```suggestion
- If `generate_file_flag` is true, the test will overwrite any saved database outputs from previous test runs. If `generate_file_flag` is false, the test will only compare the the actual database output with the expected database output and log differences. 
```

`custom_output_path`
- An optional configuration to specify a custom path where the expected database output will be stored.
If not provided, the test will use the default path defined by DEFAULT_OUTPUT_FOLDER.

`DEFAULT_OUTPUT_FOLDER`
- This constant defines the default folder where the system stores output files for the tests.
Example: "sdk_expected_db_output_files".
Modify this value in your configuration if you prefer a different default directory.

```rust
let (generate_file_flag, custom_output_path) = get_test_config();
let output_path = custom_output_path.unwrap_or_else(|| format!("{}/imported_mainnet_txns", DEFAULT_OUTPUT_FOLDER));

// Setup DB and replace as needed
let mut db = PostgresTestDatabase::new();
db.setup().await.unwrap();

let mut test_context = SdkTestContext::new(&[CONST_VARIABLE_OF_YOUR_TEST_TRANSACTION]); // Replace with your test transaction
if test_context.init_mock_grpc().await.is_err() {
    panic!("Failed to initialize mock grpc");
};

> ```

Reviewer: what does the user do with this - do we they add it to a rust test?

    let db_values = test_context
        .run(
            &processor,
            generate_file_flag,
            output_path.clone(),
            custom_file_name,
            move || {
                let mut conn = PgConnection::establish(&db_url).unwrap_or_else(|e| {
                    eprintln!("[ERROR] Failed to establish DB connection: {:?}", e);
                    panic!("Failed to establish DB connection: {:?}", e);
                });

                let db_values = match load_data(&mut conn, txn_versions.clone()) {
                    Ok(db_data) => db_data,
                    Err(e) => {
                        eprintln!("[ERROR] Failed to load data {}", e);
                        return Err(e);
                    },
                };

                if db_values.is_empty() {
                    eprintln!("[WARNING] No data found for versions: {:?}", txn_versions);
                }

                Ok(db_values)
            },
        )
```

### 6. Run the Processor Test

Once you have your test ready, run the following command to generate the expected output for validation:

```bash
cargo test sdk_tests -- generate-output
```

Arguments:

> generate-output: A custom flag to indicate that expected outputs should be generated.

Reviewer: ```suggestion
generate-output: Set this true if you want to generate or overwrite saved database output, or false if you want to compare database outputs in diff mode. 
```

> output-path: it's an optional argument to specify the output path for the db output.

Reviewer: ```suggestion
output-path: An optional argument to specify a custom output path for the db output.
```

The expected database output will be saved in the specified output_path or sdk_expected_db_output_files by default.

---

## FAQ

### What Types of Tests Does It Support?

> - Database schema output diff.

Reviewer: ```suggestion
- The testing framework allows you to write tests that compare the database outputs of processors. It helps you catch changes in database output when you're updating or developing your processor.
```

    &mut self,
    processor: &impl ProcessorTrait,
    txn_version: u64,
    generate_files: bool,             // Flag to control file generation
    output_path: String,              // Output path
    custom_file_name: Option<String>, // Custom file name
    verification_f: F,                // Verification function
) -> anyhow::Result<HashMap<String, Value>>
where
```

### How to Generate Expected DB Output?

Run the following command:

```bash
cargo test sdk_tests -- --nocapture generate-output
```

Supported Test Args:

1. `generate-output`
2. `output_path`

---

## Troubleshooting and Tips

1. **Isolate Tests**: Use Docker containers for database isolation.
2. **Handle Non-Deterministic Fields**: Use helpers like `remove_inserted_at` to clean up timestamps before validation.
3. **Enable Debugging**: Use `eprintln!` for detailed error logging.

#### How to Debug Test Failures?
run following command to get detailed logs:

```bash
cargo test sdk_tests -- --nocapture
```

## Additional Notes

- **Adapting to Other Databases**:
  - Replace PostgreSQL-specific code with relevant database code you intend to use (e.g., MySQL).
  - Update schema initialization and query methods.

> - **Docker Installation**: Follow the [Docker setup guide](https://docs.docker.com/get-docker/).

Reviewer: we should move this higher in a pre-reqs section so it's more visible

> - **Referencing Existing Tests**:

Reviewer: ```suggestion
- **References to Processor Tests**:
```
